### ISSUE TITLE
Observation.value in "ValueSet" is selecting string overload

### ISSUE THREAD

**Issue by brynrhodes** [cqframework/clinical_quality_language#564]
Given this library:

```
library TestStringInValueSet

using FHIR version '4.0.1'

include FHIRHelpers version '4.0.1'

valueset "Admit Inpatient": 'TBD'

context Patient

define Observations:
  [Observation] O
    where O.value in "Admit Inpatient"
```

This is most likely due to conversions involving primitives receiving conversion precedence over conversions involving complex types (see [Conversion Precedence](https://cql.hl7.org/03-developersguide.html#conversion-precedence)), but it is surprising behavior. Verify the translation is correct, and consider a way to address this, it leads to surprising results for a very common use case.

**Comment by brynrhodes** [comment]
Looking into this a bit more, this has also surfaced in some FHIRPath testing, where the same issue occurs, but in trying to compare to a decimal. This is a consequence of resolving choice types too aggressively, and can be handled much more effectively by allowing choice types to participate in the operator resolution process, rather than just the conversion process.

Another example of the behavior:
````
library TestFHIRPath
using FHIR version '4.0.0'
include FHIRHelpers version '4.0.0' called FHIRHelpers
parameter Observation Observation
define Test: Observation.value.value > 0
````

In this case, the resolution of the "value.value" property accessor applies to several choices, including "boolean", "decimal", "string", and "Quantity", and again the preference for simple types resolves to the decimal type, rather than the intended Quantity type. The proposed solution would allow the resolution of choice types to be carried through operator resolution, so that this would be interpreted as:

````
case
  when Observation.value is decimal then (Observation.value as decimal).value > 0
  when Observation.value is Quantity then (Observation.value as Quantity).value > 0
  else null
end
````

**Comment by brynrhodes** [comment]
@cmoesel , I'm considering this as a solution, but practically speaking, what it will mean is that operator resolutions involving choice type arguments will start to have `is` and `as` operators output in the ELM, which will effectively mean run-time JS evaluation errors when using the translator with FHIR models. There would be a workaround (using "custom casting operators"), but I'd like to get your thoughts on this proposed approach, when you get a chance.

**Comment by cmoesel** [comment]
Hi @brynrhodes.  We're actually working on implementing `Is` and `As` now, so it seems the timing is good.  Regarding the first example:
```
define Observations:
  [Observation] O
    where O.value in "Admit Inpatient"
```
Would this become:
```
define Observations:
  [Observation] O
    where FHIRHelpers.ToConcept(O.value as FHIR.CodeableConcept) in "Admit Inpatient"
```
or
```
define Observations:
  [Observation] O
    where O.value is FHIR.CodeableConcept
    and FHIRHelpers.ToConcept(O.value as FHIR.CodeableConcept) in "Admit Inpatient"
```
or something else?

**Comment by brynrhodes** [comment]
Hi @cmoesel , it will actually look something like this:

````
define Observations:
  [Observation] O
    where
      case
        when O.value is FHIR.string 
          then FHIRHelpers.ToString(O.value as FHIR.string) in "Admit Inpatient"
        when O.value is FHIR.CodeableConcept
          then FHIRHelpers.ToConcept(O.value as FHIR.CodeableConcept) in "Admit Inpatient"
        else null
      end
````
Does that help?

**Comment by cmoesel** [comment]
Yes, that does help.  Thanks.  So... this will definitely render the current `cql-execution` implementation completely broken, but as I noted, we're working on support for `Is`/`As` now.  What is your timeline?

The approach seems reasonable to me.  I suppose the only real alternative is to force users to say what they mean, right?  E.g., to require them to perform casts in order to use type-specific operators? Honestly, I'm kind of a fan of making people be explicit about some things -- because in the example above, the author probably really only _intended_ for it to work w/ `CodeableConcept`, but in our attempt to make it easier for them, we end up representing it as the broadest possible interpretation.  Same for the example before that -- the author likely had either `decimal` or `quantity` in mind when they wrote it, but in allowing them to be lazy about it, we lose the precision of their actual intent.

But I'm guessing that requiring this type of explicit indication in potentially ambiguous situations has already been deemed unacceptable to CQL authors, right?

**Comment by brynrhodes** [comment]
The timeline for this is the 1.5 translator release, which I'm hoping to wrap up as quickly as possible. Given the impact (and the fact that it's still a "correct" interpretation either way, it's just tricky to figure out what happened in some cases), would it make sense to introduce this as a translator option?
As far as being explicit, yes, you can always make it explicit (and that would still be true even with this new behavior), but the fact that you have to be explicit is really an artifact of the flexibility of the underlying model, and ideally, profile-informed authoring views of the model will eliminate the need entirely (i.e. a CodeableObservation profile would limit the available choices for value, so the choice-based resolution isn't necessary).

**Comment by cmoesel** [comment]
I don't know if it should be an option or not.  I don't want to force additional complexity on the implementation.  So, if it does make for a fair amount of complexity, I'm OK _not_ making it an option.  But if it's just a few lines or code or a simple switch statement then perhaps it makes sense for people who want to force authors to be more explicit.
